Add Rust quantum volume function

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit adds a new function quantum_volume used for generating a quantum volume model circuit. This new function is defined in Rust and multithreaded to improve the throughput of the circuit generation. This new function will eventually replace the existing QuantumVolume class as part of #13046. Since quantum volume is a circuit defined by it's structure using a generator function is inline with the goals of #13046.

Right now the performance is bottlenecked by the creation of the UnitaryGate objects as these are still defined solely in Python. We'll likely need to port the class to have a rust native representation to further speed up the construction of the circuit.

### Details and comments

mtreinish committed:
Add Rust quantum volume function

This commit adds a new function quantum_volume used for generating a
quantum volume model circuit. This new function is defined in Rust and
multithreaded to improve the throughput of the circuit generation. This
new function will eventually replace the existing QuantumVolume class as
part of #13046. Since quantum volume is a circuit defined by it's
structure using a generator function is inline with the goals of #13046.

Right now the performance is bottlenecked by the creation of the
UnitaryGate objects as these are still defined solely in Python. We'll
likely need to port the class to have a rust native representation to
further speed up the construction of the circuit.

mtreinish commented:
I ran some benchmarks comparing the construction speed of the quantum volume circuit using a logarithmic sweep from 10 to 1000 qubits (with equal depth for each point) using this function in parallel and serially (by setting `QISKIT_IN_PARALLEL=TRUE`) and comparing that against the Python `QuantumVolume` class:

![construct_time](https://github.com/user-attachments/assets/dcec3297-78dd-4e5e-aeed-06a7274059a7)

All these results were on my laptop with a AMD Ryzen 7 PRO 7840U.

There is potentially some tuning we might want to do about the number of threads and the chunk size for each parallel worker. But this seems like a reasonable balance and nothing I've changed so far appreciably changed the performance. It might be worth testing for the serial path though if doing the unitary generation all up front, like what is done in the parallel case, improves performance.

Cryoris commented on crates/accelerate/src/circuit_library/quantum_volume.rs:
```diff
@@ -0,0 +1,149 @@
+use qiskit_circuit::Qubit;
+use smallvec::smallvec;
+
+#[inline]
```
Would it make sense to link the paper to explain where the logic is coming from? https://arxiv.org/pdf/math-ph/0609050v2

mtreinish replied:
I actually based the logic on how the scipy function we were previously using worked internally. But sure I can add a link to that here: https://github.com/scipy/scipy/blob/v1.14.1/scipy/stats/_multivariate.py#L4224-L4256

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/13238/commits/69d4cfa50a25631a37caa12ab16a366ac48845b4

Cryoris commented on crates/accelerate/src/circuit_library/quantum_volume.rs:
```diff
@@ -0,0 +1,149 @@
+        let qr = z.view().into_faer_complex().qr();
+        let r = qr.compute_r().as_ref().into_ndarray_complex().to_owned();
+        let mut d = r.into_diag();
+        d.mapv_inplace(|d| d / d.norm());
```
Does this re-compute the norm of `d` in every application? If yes would it make sense to pre-compute `d.norm()`?

mtreinish replied:
This does an in-place substitution on the diagonal `d`  so that each value also named `d` (which is probably why this is confusing) is replaced with `d / d.norm()` for each element `d`. I'll rename the diagonal to `diag` or something to make it clear which is an element and which is the array.

mtreinish replied:
I am wondering if I should rewrite this using fixed size arrays or nalgebra's `Matrix4<Complex64>` to make this all faster. Right now I'm pretty confident the overhead is mostly dealing with allocations for 2x2 arrays.

mtreinish replied:
I did this here: https://github.com/Qiskit/qiskit/pull/13238/commits/32b4c1f8c32a7e531c9bc4b12288be89b3cc5b38 the code is a lot more verbose now, but I think it's a bit more explicit and it'll be a bit more optimized (although the random unitary construction is not the bottleneck).

Cryoris replied:
Nice! The bottleneck is the `UnitaryGate` creation Python-side as you wrote above, right? I think a 10x speedup (maybe more now with the explicit matrix allocation?) is already great, plus we have the circuit as function reaching the goal of #13046 😄 But if you'd like to test with pre-generating the gates I'm also happy to wait 🙂

mtreinish replied:
Yeah, we are totally bottlenecked by the `UnitaryGate` object creation, I ran a profile with the current state of the PR running the test script I was using to generate the data for the graph I shared above. In serial mode (py-spy doesn't handle the multithreading well) and the `UnitaryGate` object creation is taking ~58% of the time, while the random unitary generation is only ~30% of the runtime (almost all of that is performing the qr factorization).

I'll do a quick test of serial mode dumping the unitaries to a vec first to see if that makes a difference performance wise (I don't expect it to), and will leave a comment here with the results.

mtreinish replied:
![times](https://github.com/user-attachments/assets/b5d74a67-483a-47ad-8344-a4f21889e1bd)

I reran the numbers collecting the unitaries into a vec and then iterating off of that instead of passing the iterator directly to the circuit constructor. It looks like it's marginally faster to collect into a vec.

Cryoris commented on crates/accelerate/src/circuit_library/quantum_volume.rs:
```diff
@@ -0,0 +1,149 @@
+        let mut d = r.into_diag();
+        d.mapv_inplace(|d| d / d.norm());
+        let mut q = qr.compute_q().as_ref().into_ndarray_complex().to_owned();
+        q.axis_iter_mut(Axis(0)).for_each(|mut row| {
```
In the paper they return `q @ d_normalized @ q` -- is this missing the second `q` application here? 🤔

Cryoris replied:
Ok no, Scipy also has `q @ d_normalized` and in the equations of the paper they describe the same. Just the pseudo-code of the paper shows the additional multiplication with `q`, maybe that's strange notation or a typo. All good then 🙂

Cryoris commented on qiskit/circuit/library/quantum_volume.py:
```diff
@@ -113,3 +114,37 @@ def __init__(
+def quantum_volume(
+    num_qubits: int,
+    depth: Optional[int] = None,
+    seed: Optional[Union[int, np.random.Generator]] = None,
```
```suggestion
    depth: int | None = None,
    seed: int | np.random.Generator | None = None,
```

mtreinish committed:
Adjust type hints on python function

Co-authored-by: Julien Gacon <gaconju@gmail.com>

mtreinish committed:
Add missing __future__ import

The previous commit was relying on the behavior of the annotations
future import but neglected to add it. This commit corrects the
oversight.

mtreinish committed:
Add comment on random unitary algorithm

mtreinish committed:
Reduce allocations random_unitaries

The previous implementation had 4 heap allocations for each random
unitary constructed, this commit uses some fixed sized stack allocated
arrays and reduces that to two allocations one for q and r from the
factorization. We'll always need at least one for the `Array2` that gets
stored in each `UnitaryGate` as a numpy array. But to reduce to just
this we'll need a method of computing the QR factorization without an
allocation for the result space, nalgebtra might be a path for doing
that. While this currently isn't a bottleneck as the `UnitaryGate`
python object creation is the largest source of runtime, but assuming
that's fixed in the future this might have a larger impact.

mtreinish committed:
Merge branch 'main' into rust-qv

mtreinish committed:
Preallocate unitaries for serial path

When executing in the serial path we previously were working directly
with an iterator where the 2q unitaries we're created on the iterator
that we passed directly to circuit constructor. However testing shows
that precomputing all the unitaries into a Vec and passing the iterator
off of that to the circuit constructor is marginally but consistently
faster. So this commit pivots to using that instead.

Cryoris commented on qiskit/circuit/library/quantum_volume.py:
```diff
@@ -113,3 +116,37 @@ def __init__(
+
+    [1] A. Cross et al. Validating quantum computers using
+    randomized model circuits, Phys. Rev. A 100, 032328 (2019).
+    [`arXiv:1811.12926 <https://arxiv.org/abs/1811.12926>`_]
```
I know it's like this above, but shouldn't it be the following?
```suggestion
    `arXiv:1811.12926 <https://arxiv.org/abs/1811.12926>`__
```

mtreinish committed:
Fix determinism and error handling of of qv function

This commit fixes two issues in the reproducibility of the quantum
volume circuit. The first was the output unitary matrices for a fixed
seed would differ between the parallel and serial execution path. This
was because how the RNGs were used was different in the different code
paths. This change results in the serial path being marginally less
efficient, but it shouldn't be a big deal when compared to getting
different results in different contexts. The second was the seed usage
in parallel mode was dependent on the number of threads on the local
system. This was problematic because the exact circuit generated between
two systems would be different even with a fixed seed. This was fixed to
avoid depending on the number of threads to determine how the seeds were
used across multiple threads.

The last fix here was a change to the error handling so that the
CircuitData constructor used to create the circuit object can handle a
fallible iterator. Previously we were throwing away the python error and
panicking if the Python call to generate the UnitaryGate object raised
an error for any reason.

mtreinish committed:
Mention the new function is multithreaded in docstring

mtreinish committed:
Merge branch 'main' into rust-qv

mtreinish committed:
Update qiskit/circuit/library/quantum_volume.py

Co-authored-by: Julien Gacon <gaconju@gmail.com>

Cryoris committed:
Merge branch 'main' into rust-qv

Cryoris reviewed (approved):
LGTM and thanks for all the detailed benchmarks! 😄
